// rdme/dist/src/cmds/docs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const baseCommand_1 = __importStar(require("../../lib/baseCommand"));
const createGHA_1 = __importDefault(require("../../lib/createGHA"));
const syncDocsPath_1 = __importDefault(require("../../lib/syncDocsPath"));
const versionSelect_1 = require("../../lib/versionSelect");
class DocsCommand extends baseCommand_1.default {
    constructor() {
        super();
        this.command = 'docs';
        this.usage = 'docs <path> [options]';
        this.description =
            'Sync Markdown files to your ReadMe project as Guides. Can either be a path to a directory or a single Markdown file.';
        this.cmdCategory = baseCommand_1.CommandCategories.DOCS;
        this.hiddenArgs = ['filePath'];
        this.args = [
            this.getKeyArg(),
            this.getVersionArg(),
            {
                name: 'filePath',
                type: String,
                defaultOption: true,
            },
            this.getGitHubArg(),
            {
                name: 'dryRun',
                type: Boolean,
                description: 'Runs the command without creating/updating any docs in ReadMe. Useful for debugging.',
            },
        ];
    }
    async run(opts) {
        await super.run(opts);
        const { dryRun, filePath, key, version } = opts;
        // TODO: should we allow version selection at all here?
        // Let's revisit this once we re-evaluate our category logic in the API.
        // Ideally we should ignore this parameter entirely if the category is included.
        const selectedVersion = await (0, versionSelect_1.getProjectVersion)(version, key);
        return (0, syncDocsPath_1.default)(key, selectedVersion, this.cmdCategory, this.usage, filePath, dryRun).then(msg => (0, createGHA_1.default)(msg, this.command, this.args, { ...opts, version: selectedVersion }));
    }
}
exports.default = DocsCommand;
