// rdme/dist/src/cmds/changelogs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const baseCommand_1 = __importStar(require("../lib/baseCommand"));
const supportsGHA_1 = __importDefault(require("../lib/decorators/supportsGHA"));
const syncDocsPath_1 = __importDefault(require("../lib/syncDocsPath"));
let ChangelogsCommand = class ChangelogsCommand extends baseCommand_1.default {
    constructor() {
        super();
        this.command = 'changelogs';
        this.usage = 'changelogs <path> [options]';
        this.description =
            'Sync Markdown files to your ReadMe project as Changelog posts. Can either be a path to a directory or a single Markdown file.';
        this.cmdCategory = baseCommand_1.CommandCategories.CHANGELOGS;
        this.hiddenArgs = ['filePath'];
        this.args = [
            this.getKeyArg(),
            {
                name: 'filePath',
                type: String,
                defaultOption: true,
            },
            this.getGitHubArg(),
            {
                name: 'dryRun',
                type: Boolean,
                description: 'Runs the command without creating/updating any changelogs in ReadMe. Useful for debugging.',
            },
        ];
    }
    async run(opts) {
        await super.run(opts);
        const { dryRun, filePath, key } = opts;
        return (0, syncDocsPath_1.default)(key, undefined, this.cmdCategory, this.usage, filePath, dryRun);
    }
};
ChangelogsCommand = __decorate([
    supportsGHA_1.default
], ChangelogsCommand);
exports.default = ChangelogsCommand;
