// rdme/dist/src/cmds/whoami.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = __importDefault(require("chalk"));
const config_1 = __importDefault(require("config"));
const baseCommand_1 = __importStar(require("../lib/baseCommand"));
const getCurrentConfig_1 = __importDefault(require("../lib/getCurrentConfig"));
class WhoAmICommand extends baseCommand_1.default {
    constructor() {
        super();
        this.command = 'whoami';
        this.usage = 'whoami';
        this.description = 'Displays the current user and project authenticated with ReadMe.';
        this.cmdCategory = baseCommand_1.CommandCategories.ADMIN;
        this.args = [];
    }
    async run(opts) {
        await super.run(opts);
        const { email, project } = (0, getCurrentConfig_1.default)();
        if (!email || !project) {
            return Promise.reject(new Error(`Please login using \`${config_1.default.get('cli')} login\`.`));
        }
        return Promise.resolve(`You are currently logged in as ${chalk_1.default.green(email)} to the ${chalk_1.default.blue(project)} project.`);
    }
}
exports.default = WhoAmICommand;
